django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js

```diff
@@ -376,12 +420,110 @@
+            
+            // Check if there's a selected date (using the selected class)
+            focusTarget = cal_box.querySelector('td.selected a');
+            if (!focusTarget) {
+                // look for selected class on the link itself
+                focusTarget = cal_box.querySelector('td a.selected');
+            }
+            
+            // If no selected date found, try to focus on today's date
+            if (!focusTarget) {
+                const today = DateTimeShortcuts.now(); // Use server-time aware date
+                const todayDay = today.getDate();
+                
+                const allDateCells = cal_box.querySelectorAll('td a');
+                for (const dateCell of allDateCells) {
+                    const cellText = dateCell.textContent.trim();
+                    if (cellText === todayDay.toString()) {
+                        const cellParent = dateCell.parentElement;
+                        if (!cellParent.classList.contains('other') && !cellParent.classList.contains('noday')) {
+                            focusTarget = dateCell;
+                            break;
+                        }
+                    }
+                }
+            }
+            
+            // Final fallback: focus on first available date
+            if (!focusTarget) {
+                focusTarget = cal_box.querySelector('td a');
+            }
+            
+            // Focus on the determined target
+            if (focusTarget) {
+                focusTarget.focus();
+            } else {
+                // Ultimate fallback to first focusable element
+                const firstFocusableElement = cal_box.querySelector('a[href], button, [tabindex]:not([tabindex="-1"])');
+                if (firstFocusableElement) {
+                    firstFocusableElement.focus();
+                } else {
+                    cal_box.focus();
+                }
+            }
+
             document.addEventListener('click', DateTimeShortcuts.dismissCalendarFunc[num]);
         },
         dismissCalendar: function(num) {
-            document.getElementById(DateTimeShortcuts.calendarDivName1 + num).style.display = 'none';
+            const cal_box = document.getElementById(DateTimeShortcuts.calendarDivName1 + num);
+            cal_box.style.display = 'none';
+            
+            // Remove focus loop
+            DateTimeShortcuts.removeFocusLoop(cal_box);
+            
+            // Restore focus to the trigger button
+            if (DateTimeShortcuts.lastFocusedElement) {
+                DateTimeShortcuts.lastFocusedElement.focus();
+                DateTimeShortcuts.lastFocusedElement = null;
+            }
+            
             document.removeEventListener('click', DateTimeShortcuts.dismissCalendarFunc[num]);
         },
+        // Add focus loop to keep focus within the popup even after reaching the last focusable element
+        addFocusLoop: function(container) {
```

Comment:
I think we should try to change this to use the `inert` HTML attribute so we don’t reinvent our own focus trapping.